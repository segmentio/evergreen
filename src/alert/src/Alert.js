import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { spacing, dimensions, position, layout } from 'ui-box'
import { colors } from '../../colors'
import { Pane } from '../../layers'
import { Text } from '../../typography'
import {
  CheckCircleIcon,
  DangerIcon,
  QuestionIcon,
  WarningIcon
} from '../../icons'

const getColorForType = type => {
  switch (type) {
    case 'success':
      return colors.green['500']
    case 'question':
    default:
      return colors.blue['500']
    case 'danger':
      return colors.red['500']
    case 'warning':
      return colors.yellow['500']
  }
}

const getIconForType = type => {
  const iconProps = {
    size: 22,
    iconSize: 14,
    color: getColorForType(type)
  }

  switch (type) {
    case 'success':
      return <CheckCircleIcon {...iconProps} />
    case 'question':
    default:
      return <QuestionIcon {...iconProps} />
    case 'danger':
      return <DangerIcon {...iconProps} />
    case 'warning':
      return <WarningIcon {...iconProps} />
  }
}

const getStyle = ({ type }) => ({
  '&:before': {
    content: '""',
    width: 3,
    height: '100%',
    position: 'absolute',
    top: 0,
    left: 0,
    backgroundColor: getColorForType(type)
  }
})

const AlertAppearances = {
  default: {
    boxShadow: `inset 0 0 0 1px ${colors.neutral['20A']}`
  },
  card: {
    elevation: 1,
    borderRadius: 3
  }
}

export default class Alert extends PureComponent {
  static propTypes = {
    /**
     * Composes some Box APIs.
     */
    ...spacing.propTypes,
    ...position.propTypes,
    ...layout.propTypes,
    ...dimensions.propTypes,

    /**
     * The content of the alert. When a string is passed it is wrapped in a `<Text size={400} />` component.
     */
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),

    /**
     * The type of the alert.
     */
    type: PropTypes.oneOf([
      'default',
      'success',
      'warning',
      'danger',
      'question'
    ]),

    /**
     * The title of the alert.
     */
    title: PropTypes.node,

    /**
     * When true, show a border on the left matching the type.
     */
    hasTrim: PropTypes.bool,

    /**
     * When true, show a icon on the left matching the type,
     */
    hasIcon: PropTypes.bool,

    /**
     * The appearance of the alert.
     */
    appearance: PropTypes.oneOf(['default', 'card'])
  }

  static defaultProps = {
    type: 'default',
    hasTrim: true,
    hasIcon: true,
    appearance: 'default'
  }

  render() {
    const {
      title,
      type,
      hasTrim,
      hasIcon,
      children,
      appearance,
      ...props
    } = this.props

    const style = hasTrim && type === 'default' ? {} : getStyle({ type })

    const appearanceProps = AlertAppearances[appearance]

    return (
      <Pane
        role="alert"
        backgroundColor="white"
        overflow="hidden"
        position="relative"
        display="flex"
        paddingY={12}
        paddingX={16}
        css={style}
        {...appearanceProps}
        {...props}
      >
        {hasIcon &&
          type !== 'default' && (
            <Pane marginRight={8}>{getIconForType(type)}</Pane>
          )}
        <Pane>
          <Text
            is="h4"
            fontWeight={600}
            size={400}
            marginTop={0}
            marginBottom={0}
          >
            {title}
          </Text>
          {typeof children === 'string' ? (
            <Text size={400} color="muted">
              {children}
            </Text>
          ) : (
            children
          )}
        </Pane>
      </Pane>
    )
  }
}
