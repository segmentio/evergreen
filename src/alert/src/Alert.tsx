import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import { withTheme, Theme } from '../../theme'
import { Pane } from '../../layers'
import { Heading, Paragraph } from '../../typography'
import { IconButton } from '../../buttons'
import { Icon } from '../../icon'

type Appearance = 'default' | 'card'
type Intent = 'none' | 'success' | 'warning' | 'danger'

interface AlertProps extends React.ComponentProps<typeof Box> {
  /**
   * The appearance of the alert.
   */
  appearance: Appearance

  /**
   * The content of the alert. When a string is passed it is wrapped in a `<Text size={400} />` component.
   */
  children?: React.ReactNode

  /**
   * When true, show a icon on the left matching the type,
   */
  hasIcon?: boolean

  /**
   * When true, show a border on the left matching the type.
   */
  hasTrim?: boolean

  /**
   * The intent of the alert.
   */
  intent: Intent

  /**
   * When true, show a remove icon button.
   */
  isRemoveable?: boolean

  /**
   * Function called when the remove button is clicked.
   */
  onRemove?: () => void

  /**
   * Theme provided by ThemeProvider.
   */
  theme: Theme

  /**
   * The title of the alert.
   */
  title?: React.ReactNode
}

class Alert extends PureComponent<AlertProps> {
  static propTypes = {
    appearance: PropTypes.oneOf(['default', 'card']) as PropTypes.Validator<
      Appearance
    >,
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    hasIcon: PropTypes.bool,
    hasTrim: PropTypes.bool,
    intent: PropTypes.oneOf(['none', 'success', 'warning', 'danger'])
      .isRequired as PropTypes.Validator<Intent>,
    isRemoveable: PropTypes.bool,
    onRemove: PropTypes.func,
    theme: PropTypes.object.isRequired as PropTypes.Validator<Theme>,
    title: PropTypes.node
  }

  static defaultProps: Partial<AlertProps> = {
    appearance: 'default',
    hasIcon: true,
    hasTrim: true,
    intent: 'none',
    isRemoveable: false
  }

  getIconForIntent = intent => {
    const { theme } = this.props

    return <Icon size={14} {...theme.getIconForIntent(intent)} />
  }

  render() {
    const {
      theme,

      title,
      intent,
      hasTrim,
      hasIcon,
      children,
      appearance,
      isRemoveable,
      onRemove,
      ...props
    } = this.props

    /**
     * Note that Alert return a className and additional properties.
     */
    const { className, ...themeProps } = theme.getAlertProps({
      appearance,
      intent,
      hasTrim
    })

    return (
      <Pane
        className={className}
        role="alert"
        backgroundColor="white"
        overflow="hidden"
        position="relative"
        display="flex"
        paddingY={12}
        paddingX={16}
        {...themeProps}
        {...props}
      >
        {hasIcon && (
          <Pane
            marginRight={10}
            marginLeft={2}
            height={14}
            display="block"
            marginTop={2}
          >
            {this.getIconForIntent(intent)}
          </Pane>
        )}
        <Pane display="flex" width="100%">
          <Pane flex={1}>
            <Heading
              is="h4"
              fontWeight={600}
              size={400}
              marginTop={0}
              marginBottom={0}
            >
              {title}
            </Heading>
            {typeof children === 'string' ? (
              <Paragraph size={400} color="muted">
                {children}
              </Paragraph>
            ) : (
              children
            )}
          </Pane>
          {isRemoveable && (
            <Pane
              marginLeft={24}
              flexShrink={0}
              marginBottom={-2}
              marginTop={-2}
              marginRight={-2}
            >
              <IconButton
                icon="cross"
                appearance="minimal"
                height={24}
                onClick={onRemove}
              />
            </Pane>
          )}
        </Pane>
      </Pane>
    )
  }
}

export default withTheme(Alert)
