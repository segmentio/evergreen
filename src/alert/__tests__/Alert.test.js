import React from 'react'
import { shallow } from 'enzyme'
import render from 'react-test-renderer'

import Alert from '../src/Alert'

describe('<Alert />', () => {
  it('basic snapshot', () => {
    const component = <Alert title="A simple general message" />
    const tree = render.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('outputs title', () => {
    const component = shallow(<Alert title="Test title" />)
    expect(component.html()).toEqual(expect.stringContaining('Test title'))
  })

  it('outputs children', () => {
    const component = shallow(<Alert title="Test title">Test content</Alert>)
    expect(component.html()).toEqual(expect.stringContaining('Test content'))
  })

  it('intent snapshot', () => {
    const component = <Alert title="Test title" intent="danger" />
    const tree = render.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('appearance snapshot', () => {
    const component = <Alert title="Test title" appearance="card" />
    const tree = render.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
