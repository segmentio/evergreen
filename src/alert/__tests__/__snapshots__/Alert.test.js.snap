// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> appearance snapshot 1`] = `
<div
  className="ub-b-top_1px-solid-3366FF ub-b-rgt_1px-solid-3366FF ub-b-btm_1px-solid-3366FF ub-b-lft_1px-solid-3366FF ub-pst_relative ub-ovflw-x_hidden ub-ovflw-y_hidden ub-dspl_flex ub-pb_15px ub-pl_15px ub-pr_15px ub-pt_15px ub-bg-clr_F3F6FF ub-bblr_8px ub-bbrr_8px ub-btlr_8px ub-btrr_8px ub-color_2952CC ub-box-szg_border-box"
  role="alert"
>
  <div
    className="ub-mr_16px ub-ml_2px ub-mt_-1px ub-dspl_flex ub-algn-itms_flex-start ub-box-szg_border-box"
  >
    <svg
      className="css-1ffn9ia ub-w_16px ub-h_16px ub-box-szg_border-box"
      data-icon="info-sign"
      viewBox="0 0 16 16"
    >
      <path
        d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="ub-flx_1 ub-box-szg_border-box"
  >
    <h4
      className="ub-mt_0px ub-mb_0px ub-color_inherit ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_14px ub-ln-ht_1 ub-ltr-spc_-0-05px ub-box-szg_border-box"
    >
      Test title
    </h4>
    <div
      className="ub-box-szg_border-box"
    />
  </div>
</div>
`;

exports[`<Alert /> basic snapshot 1`] = `
<div
  className="ub-b-top_1px-solid-3366FF ub-b-rgt_1px-solid-3366FF ub-b-btm_1px-solid-3366FF ub-b-lft_1px-solid-3366FF ub-pst_relative ub-ovflw-x_hidden ub-ovflw-y_hidden ub-dspl_flex ub-pb_15px ub-pl_15px ub-pr_15px ub-pt_15px ub-bg-clr_F3F6FF ub-bblr_8px ub-bbrr_8px ub-btlr_8px ub-btrr_8px ub-color_2952CC ub-box-szg_border-box"
  role="alert"
>
  <div
    className="ub-mr_16px ub-ml_2px ub-mt_-1px ub-dspl_flex ub-algn-itms_flex-start ub-box-szg_border-box"
  >
    <svg
      className="css-1ffn9ia ub-w_16px ub-h_16px ub-box-szg_border-box"
      data-icon="info-sign"
      viewBox="0 0 16 16"
    >
      <path
        d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="ub-flx_1 ub-box-szg_border-box"
  >
    <h4
      className="ub-mt_0px ub-mb_0px ub-color_inherit ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_14px ub-ln-ht_1 ub-ltr-spc_-0-05px ub-box-szg_border-box"
    >
      A simple general message
    </h4>
    <div
      className="ub-box-szg_border-box"
    />
  </div>
</div>
`;

exports[`<Alert /> intent snapshot 1`] = `
<div
  className="ub-b-top_1px-solid-D14343 ub-b-rgt_1px-solid-D14343 ub-b-btm_1px-solid-D14343 ub-b-lft_1px-solid-D14343 ub-pst_relative ub-ovflw-x_hidden ub-ovflw-y_hidden ub-dspl_flex ub-pb_15px ub-pl_15px ub-pr_15px ub-pt_15px ub-bg-clr_FDF4F4 ub-bblr_8px ub-bbrr_8px ub-btlr_8px ub-btrr_8px ub-color_A73636 ub-box-szg_border-box"
  role="alert"
>
  <div
    className="ub-mr_16px ub-ml_2px ub-mt_-1px ub-dspl_flex ub-algn-itms_flex-start ub-box-szg_border-box"
  >
    <svg
      className="css-tn1wco ub-w_16px ub-h_16px ub-box-szg_border-box"
      data-icon="error"
      viewBox="0 0 16 16"
    >
      <path
        d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm1 13h-2v-2h2v2zm0-3h-2v-7h2v7z"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="ub-flx_1 ub-box-szg_border-box"
  >
    <h4
      className="ub-mt_0px ub-mb_0px ub-color_inherit ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_14px ub-ln-ht_1 ub-ltr-spc_-0-05px ub-box-szg_border-box"
    >
      Test title
    </h4>
    <div
      className="ub-box-szg_border-box"
    />
  </div>
</div>
`;
